### PR TITLE
Thomas/xml prepare for merging

### PR DESCRIPTION
**Pull request by Thomas-Ulrich**
With this pull request, I hope to answer the 3 years old review from Sebastian R.
The aim is to merge to xml branch to master.

> SR: PUMGen should be a tool that converts and creates meshes in SeisSol's mesh format from various sources. Some of your changes should be in a separate tool (e.g. exporting the mesh through the Simmoduler API) or implemented in a more general way (e.g. the mesh analysis).

>>me: you mean the AR histograms? 

>SR: I was referring to "analyse_mesh". I have a minimalistic analysis in the main function: https://github.com/SeisSol/PUMGen/blob/xml/src/pumgen.cpp#L202
The main difference, it works with any input format. You can get the min insphere not only for SimModeler meshes but also when converting a netCDF or Gambit files. Can you rewrite the analysis to work with the PUMI/APF API? 
>> Me: Aspect ratio for PUMI/APF API:It seems they have a 'measureTetQuality'
>> function, I can test it is in aggreement with the AR:
>> https://github.com/SCOREC/core/blob/61f610229a05583fcf936973e3f5042d7c24f072/ma/maQuality.cc#L139 

--> This I did not change. I think this is rather a detail and so far we only have one mesh generator within pumgen.

>> Me: Setting up the xml file without opening the GUI (could be useful for
>> very large geometry files, and/or when working remotely) was made easier
>> by the use of the --probeface, written by Johannes K.
>> I suppose it could be useful, but I never used this function so far.

> SR: This function should probably be in a different tool. 

>>"exporting the mesh through the Simmoduler API":
>>(it is useful when you have a very large topographic file (several gb),
and want to remesh it. you dont really need the GUI for that, and it may
lag a lot because of the large geometry). 

> SR: This function should probably be in a different tool. 

--> I removed the part to `export the mesh` and `probeface` functions 

>SR: The XML file is a nice feature but it is probably hard to maintain. Moreover, at the moment your code seems to be rather a prototype than a production ready version. I see a lot of debug output and copy&paste code.
>Currently, there is a lot of duplicate code to parse the options an call the corresponding SimModeler functions. Duplicate code parts usually make it hard to maintain a project. 

--> I cleaned the MeshAttribute class and Sebastian W cleaned the AnalysesAttribute class previously.

>SR: In addition, you have to test the code with native models. They have to be handled a bit differently.
>SimModeler supports discrete and native models. In the discrete models, the geometry is represented as a mesh. In native models, the geometry is stored in some kind of CAD format. The Landers setup is a native model which uses the Parasolid kernel from SimModeler. Loading native models works slightly different because you have a CAD file in addition to the smd file. PUMGen has a few if-else statements to work with both formats. 

I did not fix that as nobody uses native models anymore, and I don't have such a file to test

### CODE REVIEW

File: src/input/MeshAttributes.cpp [hunk 1]
```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
...
-         surfaceSmoothingLevel<<" "<< surfaceSmoothingType<<" "<< surfaceFaceRotationLimit <<" "<<surfaceSnap;
+    retrieve_attribute_and_update_class_member(&doc, "SurfaceMeshing", "DiscreteAngle", &surfaceFaceRotationLimit);
+    retrieve_attribute_and_update_class_member(&doc, "SurfaceMeshing", "SmoothingLevel", &surfaceSmoothingLevel);
+    retrieve_attribute_and_update_class_member(&doc, "SurfaceMeshing", "Snap", &surfaceSnap);
+    if (auto pRoot = doc.FirstChildElement("SurfaceMeshing")) {
+        std::string sSmoothingType = pRoot->Attribute("SmoothingType");
+        if (sSmoothingType.compare("Laplacian")==0) {
+           surfaceSmoothingType=0;
+        } else if (sSmoothingType.compare("Gradient")==0) {
+           surfaceSmoothingType=1;
```
Comment (sebwolf-de) [inline src/input/MeshAttributes.cpp]: You could use a class enum for surfaceSmoothingType.

File: src/input/MeshAttributes.cpp [hunk 2]
```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
...
+    if (auto pRoot = doc.FirstChildElement("VolumeMeshing")) {
+        std::string sSmoothingType = pRoot->Attribute("SmoothingType");
+        if (sSmoothingType.compare("Laplacian")==0) {
+           volumeSmoothingType=0;
+        } else if (sSmoothingType.compare("Gradient")==0) {
+           volumeSmoothingType=1;
+        } else {
+           logError() << "Unrecognised volumeSmoothingType (Laplacian or Gradient)" << sSmoothingType;
+        }
+        logInfo(PMU_rank()) << "volume smoothing option: volumeSmoothingLevel volumrSmoothingType"<<
```
Comment (sebwolf-de) [inline src/input/MeshAttributes.cpp]: Typo, although it was in the master branch before.

File: src/input/MeshAttributes.cpp [hunk 3]
```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
...
 
 
-void MeshAttributes::set_area_AspectRatio() {
-    std::string sval;
-    XMLElement* pRoot;
-    pRoot = doc.FirstChildElement("area_AspectRatio");
-    if(pRoot) {
-       sval = pRoot->Attribute("value");
-       area_AspectRatio =  atof(sval.c_str());
+void MeshAttributes::retrieve_attribute_and_update_class_member(XMLNode* element, const char* sElementName, const char* sAttributeName, double* MeshAttributesMember) {
```
Comment (sebwolf-de) [inline src/input/MeshAttributes.cpp]: You could make `element` a reference. Also https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-logical this rule states, that a function should not have two logical purposes.
What about e.g. `gradation = retrieve_attribute(doc, "gradation", "value")`

File: src/input/MeshAttributes.cpp [hunk 4]
```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
...
+    if (auto child = element->FirstChildElement(sElementName)) {
```
Comment (sebwolf-de) [inline src/input/MeshAttributes.cpp]: What do we do, if element has no child with name `sElementName`? Is there some kind of error detection?

File: src/input/MeshAttributes.cpp [hunk 5]
```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
...
+        *MeshAttributesMember = std::atof(child->Attribute(sAttributeName));
```
Comment (sebwolf-de) [inline src/input/MeshAttributes.cpp]: Same here, what do we do if child does not have an Attribute `sAttributeName`?

File: src/input/MeshAttributes.cpp [hunk 6]
```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
...
 }
-
-void MeshAttributes::set_vol_AspectRatio() {
-    std::string sval;
-    XMLElement* pRoot;
-    pRoot = doc.FirstChildElement("vol_AspectRatio");
-    if(pRoot) {
-       sval = pRoot->Attribute("value");
-       vol_AspectRatio =  atof(sval.c_str());
+void MeshAttributes::retrieve_attribute_and_update_class_member(XMLNode* element, const char* sElementName, const char* sAttributeName, int* MeshAttributesMember) {
```
Comment (sebwolf-de) [inline src/input/MeshAttributes.cpp]: Same as above

**Review by sebwolf-de — COMMENTED** [review]
Thanks for the PR. Is SR still available?

**Comment by sebwolf-de** [comment]
> --> This I did not change. I think this is rather a detail and so far we only have one mesh generator within pumgen.

I guess here we have to make a decision where to head with PUMGen. Currently it is used a a converter (gambit -> hdf5) and as glue code to SimModeler as well. With the xml functionality we make it harder to include another mesher later, for example if we want to include gmsh directly into PUMGen, we would need to map the xml input to the gmsh API and so forth.

File: src/input/MeshAttributes.cpp (hunk 1 above)
Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp]: How would you do that?
https://stackoverflow.com/questions/43510615/enum-class-of-type-string-in-c

File: src/input/MeshAttributes.cpp (hunk 3 above)
Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp]: if you do ```gradation = retrieve_attribute(doc, "gradation", "value")```, then gradation get updated even if element has no child with name ```sElementName```, that why I merged these 2 steps in one function.

File: src/input/MeshAttributes.cpp (hunk 4 above)
Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp]: In that case, the condition is false, and the parameter does not get updated?
Do you suggest adding a else logError(*) ?

File: src/input/MeshAttributes.cpp (hunk 1 above)
Comment (sebwolf-de) [reply src/input/MeshAttributes.cpp]: Have somewhere in the Header:
```
enum class MeshAttributes::SmoothingType {
  Laplacian,
  Gradient
};
```
And then 
```
if (sSmoothingType.compare("Laplacian")==0) {
           surfaceSmoothingType=SmoothingType::Laplacian;
```

Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp]: ok thx.

File: src/input/MeshAttributes.cpp (hunk 4 above)
Comment (sebwolf-de) [reply src/input/MeshAttributes.cpp]: IIRC it is valid to not give all elements, but if an element is present, the attribute has to be there to make sense right?
So having no `gradation` tag is valid, but having a `<gradation>` tag without a `value` attribute is invalid, right?

File: src/input/MeshAttributes.cpp (hunk 3 above)
Comment (sebwolf-de) [reply src/input/MeshAttributes.cpp]: If you call it `update_attribute_from_xml` there's no and in the function name any more, so automatic semantic checkers don't see this as an issue any more ;-)

Comment (sebwolf-de) [reply src/input/MeshAttributes.cpp]:
```
void MeshAttributes::update_attribute_from_xml(XMLNode& element, const char* sElementName, const char* sAttributeName, double& MeshAttributesMember) {
    if (auto child = element->FirstChildElement(sElementName)) {
        const char* attr = child->Attribute(sAttributeName);
        if (attr != null) {
            MeshAttributesMember = std::atof(attr);
        } else {
            logError() << XML Tag " << sElementName << " has no attribute " << sAttributeName;
        }
    }
}
```

Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp:133]: Thx. But actually, after some thoughts, the attr==null should not raise an error.
If the attribute is not specified it is not an error but a choice.
(e.g. in the GUI, one may choose to set a refinement zone (e.g. cube), or not).

**Comment by Thomas-Ulrich** [comment]
> > --> This I did not change. I think this is rather a detail and so far we only have one mesh generator within pumgen.
> 
> I guess here we have to make a decision where to head with PUMGen. Currently it is used a a converter (gambit -> hdf5) and as glue code to SimModeler as well. With the xml functionality we make it harder to include another mesher later, for example if we want to include gmsh directly into PUMGen, we would need to map the xml input to the gmsh API and so forth.

I think it is fine if the xml feature is only supported by SimModeler, if it does not break the functionality of pumgen for other mesh generators. Then we would only need to check that the mesh generator used is SimModeler when using xml option and if not raise an error.

File: src/input/MeshAttributes.cpp (hunk 4 above)
Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp:134]: ha yes exactly. That what I wrote in a comment above. I will see if it is easy to implement that.

File: src/input/MeshAttributes.cpp (hunk 3 above)
Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp]: ok that's fine!

**Comment by sebwolf-de** [comment]
> I think it is fine if the xml feature is only supported by SimModeler, if it does not break the functionality of pumgen for other mesh generators. Then we would only need to check that the mesh generator used is SimModeler when using xml option and if not raise an error.

Agree.
Does anyone use a workflow other than gmsh->gambit->pumgen or pumgen+simmetrix?

**Comment by Thomas-Ulrich** [comment]
No.

File: src/input/MeshAttributes.cpp (hunk 1 above)
Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp]: Hi, actually I don't really understand how to use that.
surfaceSmoothingType has to be an int not an enum, e.g., if I do 

```
SmoothingType surfaceSmoothingType = SmoothingType::Laplacian;
```
Then I get:
```
/hppfs/work/pr45fi/di73yeq4/myLibs/PUMGen/src/input/SimModSuite.h(158): error: argument of type "MeshAttributes::SmoothingType" is incompatible with parameter of type "int"
                     VolumeMesher_setSmoothType(volumeMesher, MeshAtt.volumeSmoothingType);
```

but then if I try:

```
using namespace tinyxml2;
class MeshAttributes {
  public:
    enum class SmoothingType: int {
      Laplacian = 0,
      Gradient = 1
    };
    int surfaceSmoothingLevel=2;
    int surfaceSmoothingType=1;
```
I get:
```
/hppfs/work/pr45fi/di73yeq4/myLibs/PUMGen/src/input/MeshAttributes.cpp(78): error: a value of type "MeshAttributes::SmoothingType" cannot be assigned to an entity of type "int"
             volumeSmoothingType = SmoothingType::Laplacian;
                                 ^
```

Comment (sebwolf-de) [reply src/input/MeshAttributes.cpp]: Try to change `SimModSuite.h(158)` to `VolumeMesher_setSmoothType(volumeMesher, static_cast<int>(MeshAtt.volumeSmoothingType));`
Why so complicated you might ask? Using the Simmetrix conventions (Laplacian=0, Gradient=1) everywhere in the code, makes it hard to maintain and error prone. If you are programming the xml reader, you always have to keep in mind, what their convention was. With the strongly typed enum, we just convert to int when we call the Simmetrix library acoording to their conventions.

Comment (Thomas-Ulrich) [reply src/input/MeshAttributes.cpp]: ok!

**Comment by Thomas-Ulrich** [comment]
I ve compared master and this branch and realized there is also the loadSTL function which I introduced that must be removed.
This was not working and should be part of a different tool anyway.

**Review by sebwolf-de — APPROVED** [review]

**Pull request merged**